nntrainer/src/neuralnet.cpp

```diff
@@ -213,252 +150,128 @@ int NeuralNetwork::loadFromConfig() {
 
-      addLayer(flatten_layer);
-    } break;
+  if (ini_file.empty()) {
+    ml_loge("Error: Configuration File is not defined");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
 
-    case LAYER_FC: {
-      WeightDecayParam weight_decay;
-      std::shared_ptr<FullyConnectedLayer> fc_layer =
-        std::make_shared<FullyConnectedLayer>();
+  /** Parse ini file */
+  ini = iniparser_load(ini_file.c_str());
+  if (ini == NULL) {
+    ml_loge("Error: cannot parse file: %s\n", ini_file.c_str());
+    return ML_ERROR_INVALID_PARAMETER;
+  }
 
-      status = fc_layer->setName(layer_name);
-      NN_INI_RETURN_STATUS();
+  /** Get number of sections in the file */
+  num_ini_sec = iniparser_getnsec(ini);
+  if (num_ini_sec < 0) {
+    ml_loge("Error: invalid number of sections.");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
 
-      std::string input_shape_str = iniparser_getstring(
-        ini, (layer_name + ":Input_Shape").c_str(), unknown);
-
-      if (input_shape_str.compare("Unknown") != 0) {
-        TensorDim d;
-        d.setTensorDim(input_shape_str);
-        fc_layer->setInputDimension(d);
-      } else if (section_names_iter == section_names.begin()) {
-        ml_loge("Error: %s layer input shape not specified.",
-                layer_name.c_str());
-        status = ML_ERROR_INVALID_PARAMETER;
-        NN_INI_RETURN_STATUS();
-      }
+  if (iniparser_find_entry(ini, "network") == 0) {
+    ml_loge("there is no [network] section in given ini file");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
 
-      fc_layer->setUnit(static_cast<unsigned int>(
-        iniparser_getint(ini, (layer_name + ":Unit").c_str(), 0)));
+  /** Get all the section names */
+  for (int idx = 0; idx < num_ini_sec; ++idx) {
+    const char *sec_name = iniparser_getsecname(ini, idx);
 
-      fc_layer->setBiasZero(b_zero);
-      fc_layer->setWeightInit((WeightIniType)parseType(
-        iniparser_getstring(ini, (layer_name + ":WeightIni").c_str(),
-                            "xavier_uniform"),
-        TOKEN_WEIGHTINI));
+    if (!sec_name) {
+      ml_loge("Error: Unable to retrieve section names from ini.");
+      return ML_ERROR_INVALID_PARAMETER;
+    }
 
-      status = parseWeightDecay(ini, layer_name, weight_decay);
-      NN_INI_RETURN_STATUS();
+    if (strncasecmp(network_str, sec_name, network_len) == 0) {
+      status = loadNetworkConfig((void *)ini);
+      NN_RETURN_STATUS();
+      continue;
+    }
 
-      fc_layer->setWeightDecay(weight_decay);
+    if (strncasecmp(dataset_str, sec_name, dataset_len) == 0) {
+      status = loadDatasetConfig((void *)ini);
+      NN_RETURN_STATUS();
+      continue;
+    }
 
-      addLayer(fc_layer);
-    } break;
-    case LAYER_BN: {
-      std::shared_ptr<BatchNormalizationLayer> bn_layer =
-        std::make_shared<BatchNormalizationLayer>();
+    /** Parse all the layers defined as sections in order */
+    ml_logi("parsing layers properties... undefined properties will be ignored "
```

Comment:
`undefined properties will be filled with default values.`??
Also is this needed for every layer? Probably giving this out once should be enough.

Reply:
moved location and clarified the description as
```cpp
  ml_logi(
    "parsing ini... invalid properties does not cause error, rather be ignored "
    "only invalid value for valid property will abort the process");
```